File: core/src/main/java/hunternif/voxarch/plan/Gate.java
package hunternif.voxarch.plan;

import hunternif.voxarch.vector.Vec2;
import hunternif.voxarch.vector.Vec3;

/**
 * A gate between 2 {@link Room}s, can be vertical or horizontal. A gate
 * generator should clear a passage and may want to create some sort arch
 * over it.
 * The y coordinate should be at room's origin.
 * @author Hunternif
 */
public class Gate extends Node {
	public static enum Orientation {
		VERTICAL, HORIZONTAL;
	}

	private final Room room1, room2;
	
	/** For vertical gates it is (width, height), for horizontal (width, length). */
	private final Vec2 size;
	
	private final Orientation orientation;
	
	/**
	 * For a horizontal gate, the origin point is in the middle of the gate at
	 * floor level. For a vertical gate, it is in the actual center of it.
	 * The vector arguments will be cloned.
	 */
	public Gate(Room parent, Room room1, Room room2, Vec3 origin, Vec2 size, Orientation orientation, double rotationY) {
		super(parent, origin, rotationY);
		this.room1 = room1;
		this.room2 = room2;
		this.size = new Vec2(size);
		this.orientation = orientation;
	}

	public Vec2 getSize() {
		return size;
	}
	
	public Room getRoom1() {
		return room1;
	}

	public Room getRoom2() {
		return room2;
	}

	protected Orientation getOrientation() {
		return orientation;
	}
	public boolean isVertical() {
		return orientation == Orientation.VERTICAL;
	}
	public boolean isHorizontal() {
		return orientation == Orientation.HORIZONTAL;
	}
}

File: core/src/main/kotlin/hunternif/voxarch/plan/Gate.kt
package hunternif.voxarch.plan

import hunternif.voxarch.vector.Matrix4
import hunternif.voxarch.vector.Vec2
import hunternif.voxarch.vector.Vec3
import hunternif.voxarch.vector.Vec4

/**
 * ```
 * Y
 *  +--------------------> X (East)
 *  |           inside
 *  |  origin == gate ==
 *  |           outside
 *  V
 *  Z
 * ```
 * A horizontal gate between two [Room]s.
 * The gate's generator should clear the passage and may add decorations.
 * @param origin the lower-left point, will be copied
 * @param sizeXY X=width, Y=height
 */
open class Gate(
    origin: Vec3,
    sizeXY: Vec2
) : Node(origin) {

    init {
        size.set(sizeXY.x, sizeXY.y, 0.0)
    }

    val sizeXY: Vec2 get() = Vec2(size.x, size.y)

    /** Center at floor level, relative to parent's origin. For legacy tests. */
    val center: Vec3
        get() = origin.add(
            Matrix4.rotationY(rotationY)
                .multiplyLocal(
                    Vec4(sizeXY.x/2, 0.0, 0.0, 1.0)
                ).let { Vec3.from(it) }
        )

    /**
     * legacy constructor
     * @param origin center at floor level
     */
    constructor(
        parent: Node?,
        room1: Node,
        room2: Node,
        origin: Vec3,
        sizeXY: Vec2,
        rotationY: Double
    ): this(
        // move origin from center to corner
        origin.add(
            Matrix4.rotationY(rotationY)
                .multiplyLocal(
                    Vec4(-sizeXY.x/2, 0.0, 0.0, 1.0)
                ).let { Vec3.from(it) }
        ),
        sizeXY
    ) {
        this.parent = parent
        this.rotationY = rotationY
    }

    constructor(): this(Vec3.ZERO, Vec2.ZERO)
}
